File: composeApp/src/androidMain/kotlin/com/humblesolutions/indsphinx/viewmodel/QuestionNotificationViewModel.kt
package com.humblesolutions.indsphinx.viewmodel

import android.app.Application
import android.util.Log
import android.widget.Toast
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.CreationExtras
import com.humblesolutions.indsphinx.model.QuestionNotification
import com.humblesolutions.indsphinx.repository.AndroidAuthRepository
import com.humblesolutions.indsphinx.repository.BackendQuestionNotificationRepository
import com.humblesolutions.indsphinx.usecase.SubmitQuestionNotificationResponseUseCase
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

sealed class QuestionNotificationUiState {
    object Loading : QuestionNotificationUiState()
    data class AlreadyResponded(val qn: QuestionNotification) : QuestionNotificationUiState()
    data class Ready(
        val qn: QuestionNotification,
        val selectedOptions: Set<String>,
        val textInput: String,
        val isSubmitting: Boolean,
        val isDeadlinePassed: Boolean
    ) : QuestionNotificationUiState()
    object Submitted : QuestionNotificationUiState()
    data class Error(val message: String) : QuestionNotificationUiState()
}

class QuestionNotificationViewModel(
    application: Application,
    private val qnId: String
) : AndroidViewModel(application) {

    private val authRepo = AndroidAuthRepository()
    private val repo = BackendQuestionNotificationRepository()
    private val submitUseCase = SubmitQuestionNotificationResponseUseCase(repo)

    private val _uiState = MutableStateFlow<QuestionNotificationUiState>(QuestionNotificationUiState.Loading)
    val uiState: StateFlow<QuestionNotificationUiState> = _uiState.asStateFlow()

    private var responseListenerJob: Job? = null

    init {
        loadAndObserve()
    }

    private fun loadAndObserve() {
        viewModelScope.launch {
            try {
                val uid = authRepo.getCurrentUser()?.uid ?: run {
                    _uiState.value = QuestionNotificationUiState.Error("Not signed in.")
                    return@launch
                }
                val qn = repo.fetchQuestionNotification(qnId) ?: run {
                    _uiState.value = QuestionNotificationUiState.Error("Question not found.")
                    return@launch
                }
                if (!qn.targetUserIds.contains(uid)) {
                    _uiState.value = QuestionNotificationUiState.Error("You are not a recipient of this question.")
                    return@launch
                }
                val isDeadlinePassed = qn.responseDeadline != null &&
                    qn.responseDeadline!! < System.currentTimeMillis()

                responseListenerJob = viewModelScope.launch {
                    repo.observeResponseExists(qnId, uid).collect { exists ->
                        val current = _uiState.value
                        if (current is QuestionNotificationUiState.Submitted) return@collect
                        _uiState.value = if (exists) {
                            QuestionNotificationUiState.AlreadyResponded(qn)
                        } else {
                            if (current is QuestionNotificationUiState.Ready) current
                            else QuestionNotificationUiState.Ready(
                                qn = qn,
                                selectedOptions = emptySet(),
                                textInput = "",
                                isSubmitting = false,
                                isDeadlinePassed = isDeadlinePassed
                            )
                        }
                    }
                }
            } catch (e: Exception) {
                _uiState.value = QuestionNotificationUiState.Error(e.message ?: "Failed to load question.")
            }
        }
    }

    fun toggleOption(option: String) {
        val state = _uiState.value as? QuestionNotificationUiState.Ready ?: return
        val updated = if (state.qn.allowMultipleChoices) {
            if (option in state.selectedOptions) state.selectedOptions - option
            else state.selectedOptions + option
        } else {
            setOf(option)
        }
        _uiState.value = state.copy(selectedOptions = updated)
    }

    fun setTextInput(text: String) {
        val state = _uiState.value as? QuestionNotificationUiState.Ready ?: return
        _uiState.value = state.copy(textInput = text)
    }

    fun submit(displayName: String, flatNo: String, recipientType: String) {
        val state = _uiState.value as? QuestionNotificationUiState.Ready ?: return
        val uid = authRepo.getCurrentUser()?.uid
        if (uid == null) {
            Log.w(TAG, "submit: aborted — no signed-in user")
            return
        }
        if (state.isSubmitting || state.isDeadlinePassed) {
            Log.w(TAG, "submit: aborted — isSubmitting=${state.isSubmitting} isDeadlinePassed=${state.isDeadlinePassed}")
            return
        }
        Log.d(TAG, "submit: starting qnId=$qnId uid=$uid options=${state.selectedOptions.size} hasText=${state.textInput.isNotBlank()}")
        _uiState.value = state.copy(isSubmitting = true)
        viewModelScope.launch {
            try {
                submitUseCase.execute(
                    qnId = qnId,
                    uid = uid,
                    displayName = displayName,
                    flatNo = flatNo,
                    recipientType = recipientType,
                    selectedOptions = state.selectedOptions.toList(),
                    textResponse = state.textInput
                )
                responseListenerJob?.cancel()
                Log.d(TAG, "submit: success qnId=$qnId")
                _uiState.value = QuestionNotificationUiState.Submitted
            } catch (e: Exception) {
                Log.e(TAG, "submit: FAILED qnId=$qnId — ${e.message}", e)
                Toast.makeText(
                    getApplication(),
                    "Could not submit response: ${e.message ?: "unknown error"}",
                    Toast.LENGTH_LONG,
                ).show()
                _uiState.value = state.copy(isSubmitting = false)
            }
        }
    }

    override fun onCleared() {
        super.onCleared()
        responseListenerJob?.cancel()
    }

    companion object {
        private const val TAG = "QuestionNotifFlow"

        fun factory(qnId: String) = object : ViewModelProvider.Factory {
            @Suppress("UNCHECKED_CAST")
            override fun <T : ViewModel> create(modelClass: Class<T>, extras: CreationExtras): T {
                val app = extras[ViewModelProvider.AndroidViewModelFactory.APPLICATION_KEY]!!
                return QuestionNotificationViewModel(app, qnId) as T
            }
        }
    }
}

File: iosApp/iosApp/QuestionNotificationViewModel.swift
import Foundation
import FirebaseAuth
import FirebaseFirestore

@MainActor
class QuestionNotificationViewModel: ObservableObject {

    enum State {
        case loading
        case alreadyResponded(QuestionNotification)
        case ready(QuestionNotification)
        case error(String)
    }

    private let qnId: String
    private let repo = BackendQuestionNotificationRepository()

    @Published var state: State = .loading
    @Published var selectedOptions: Set<String> = []
    @Published var textInput: String = ""
    @Published var isSubmitting: Bool = false
    @Published var isDeadlinePassed: Bool = false
    @Published var submitError: String? = nil

    private var responseListener: ListenerRegistration?

    init(qnId: String) {
        self.qnId = qnId
        Task { await loadAndObserve() }
    }

    deinit {
        responseListener?.remove()
    }

    private func loadAndObserve() async {
        guard let uid = Auth.auth().currentUser?.uid else {
            state = .error("Not signed in.")
            return
        }
        do {
            guard let qn = try await repo.fetchQuestionNotification(qnId: qnId) else {
                state = .error("Question not found.")
                return
            }
            guard qn.targetUserIds.contains(uid) else {
                state = .error("You are not a recipient of this question.")
                return
            }
            if let deadline = qn.responseDeadline, deadline < Date() {
                isDeadlinePassed = true
            }

            responseListener = repo.observeResponseExists(qnId: qnId, uid: uid) { [weak self] exists in
                Task { @MainActor in
                    guard let self else { return }
                    if case .alreadyResponded = self.state { return }
                    if exists {
                        self.state = .alreadyResponded(qn)
                    } else if case .ready = self.state {
                        // Preserve in-progress user input
                    } else {
                        self.state = .ready(qn)
                    }
                }
            }
        } catch {
            state = .error(error.localizedDescription)
        }
    }

    func toggleOption(_ option: String) {
        guard case .ready(let qn) = state else { return }
        if qn.allowMultipleChoices {
            if selectedOptions.contains(option) { selectedOptions.remove(option) }
            else { selectedOptions.insert(option) }
        } else {
            selectedOptions = [option]
        }
    }

    func submit(displayName: String, flatNo: String, recipientType: String) {
        guard case .ready(let qn) = state else {
            print("[QuestionNotifFlow] submit: aborted — not in ready state")
            return
        }
        guard let uid = Auth.auth().currentUser?.uid else {
            print("[QuestionNotifFlow] submit: aborted — no signed-in user")
            return
        }
        guard !isSubmitting, !isDeadlinePassed else {
            print("[QuestionNotifFlow] submit: aborted — isSubmitting=\(isSubmitting) isDeadlinePassed=\(isDeadlinePassed)")
            return
        }
        print("[QuestionNotifFlow] submit: starting qnId=\(qnId) uid=\(uid) options=\(selectedOptions.count) hasText=\(!textInput.isEmpty)")
        isSubmitting = true
        submitError = nil
        Task {
            do {
                try await repo.submitResponse(
                    qnId: qnId,
                    uid: uid,
                    displayName: displayName,
                    flatNo: flatNo,
                    recipientType: recipientType,
                    selectedOptions: Array(selectedOptions),
                    textResponse: textInput
                )
                responseListener?.remove()
                responseListener = nil
                print("[QuestionNotifFlow] submit: success qnId=\(qnId)")
                state = .alreadyResponded(qn)
            } catch {
                print("[QuestionNotifFlow] submit: FAILED qnId=\(qnId) error=\(error)")
                submitError = error.localizedDescription
                isSubmitting = false
            }
        }
    }
}
